[JIT] canonicalize aten::rsub

mcr229 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65014

Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

Commit:
[JIT] canonicalize aten::rsub



Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

mcr229 commented:
https://www.internalfb.com/intern/testinfra/testconsole/testrun/4222124719366029/

eellison commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
+// Normalize rsub such that `rsub(x,y) = sub(x,y)`
+bool normalizeRSub(graph_node_list_iterator& iter) {
+  ArrayRef<Value*> args = iter->inputs();
+  if (iter->kind() == aten::rsub) {
```
we all overloads for any particular symbol, so it's possible unlikely that there are other `aten::rsub` operators with different typed inputs for which this substitution isnt valid. just to be safe and future proof we should check the full schema here. you can use the `node->matches("rsub.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor")` api. 

See https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4880, for this operator what we're doing isnt valid

eellison commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
 
+// Normalize rsub such that `rsub(x,y) = sub(x,y)`
+bool normalizeRSub(graph_node_list_iterator& iter) {
+  ArrayRef<Value*> args = iter->inputs();
```
nit: move `args` into the if block

eellison commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -20,6 +20,19 @@ bool normalizeOpAliases(graph_node_list_iterator& iter) {
+bool normalizeRSub(graph_node_list_iterator& iter) {
+  ArrayRef<Value*> args = iter->inputs();
+  if (iter->kind() == aten::rsub) {
+    iter->replaceWithNewSymbol(aten::sub);
```
`replaceWithNewSymbol` returns the updated node, so the replacements we're doing in the following lines will not affect it but the existing node we're destroying

eellison reviewed (commented):
Looks good!! a few comments from me. test ?

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

mcr229 commented:
I tried running the test through buck:
 `buck test @mode/dev caffe2/test:jit -- 'test_normalized_rsub \(test_peephole\.TestPeephole\)'`
and it seemed passes, but i don't think it was doing anything, so I will try testing by building and running 
`python3 test/test_jit.py`

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

mcr229 commented:
Ran tests

`(pytorch) [maxren@devvm3115.atn0 ~/pytorch] python3 test/test_jit.py TestPeephole
CUDA not available, skipping tests
monkeytype is not installed. Skipping tests for Profile-Directed Typing
........s......................
----------------------------------------------------------------------
Ran 31 tests in 0.393s

OK (skipped=1)
(pytorch) [maxren@devvm3115.atn0 ~/pytorch] python3 test/test_jit.py TestPeephole.test_normalized_rsub
CUDA not available, skipping tests
monkeytype is not installed. Skipping tests for Profile-Directed Typing
.
----------------------------------------------------------------------
Ran 1 test in 0.015s

OK
`

eellison commented on test/jit/test_peephole.py:
```diff
@@ -219,6 +219,16 @@ def forward(self, x):
             FileCheck().check("conv").check("dim").run(conv_dim.graph)
 
+    def test_normalized_rsub(self):
+        def convertible_rsub(x: int, y: int):
```
i think we should be checking tensor inputs here

eellison reviewed (commented):
Nice! just one small comment

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

mcr229 commented on torch/csrc/jit/passes/normalize_ops.cpp:
```diff
@@ -107,6 +126,7 @@ const std::unordered_map<Symbol, Symbol>& getOperatorAliasMap() {
       {aten::true_divide_, aten::div_},
       {aten::concat, aten::cat},
       {aten::row_stack, aten::vstack},
+      {aten::rsub, aten::sub},
```
the fix for this test
![image](https://user-images.githubusercontent.com/40742183/133736305-f8f44590-579e-4137-8149-ed3e9f0becd0.png)
seemed to be that rsub could not be found within this map, so adding it to the map and making the appropriate changes seemed to make the test pass

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

Commit:
Update on "[JIT] canonicalize aten::rsub"


Differential Revision: [D30941389](https://our.internmc.facebook.com/intern/diff/D30941389/)

[ghstack-poisoned]

eellison commented on test/test_ops.py:
```diff
@@ -775,7 +775,7 @@ def get_sample():
                     if dtype == torch.float32:
                         # TODO: no reason why we cant run this with tracing graph
-                        if support_script:
+                        if support_script and op.name != "rsub":
```
@alanwaketan this is an example of an annoying test quirk that your PR will fix

eellison reviewed (approved):
Awesome!! 🚢 🚢 🚢
